presto-parquet/src/main/java/com/facebook/presto/parquet/cache/MetadataReader.java

```diff
@@ -366,17 +373,46 @@ private static void readTypeSchema(Types.GroupBuilder<?> builder, Iterator<Schem
                 }
                 typeBuilder = primitiveBuilder;
             }
-
             if (element.isSetConverted_type()) {
                 typeBuilder.as(getOriginalType(element.converted_type));
             }
+            if (element.isSetLogicalType()) {
+                LogicalType type = element.getLogicalType();
+                if (element.getLogicalType().isSetTIME()) {
+                    TimeType time = type.getTIME();
+                    typeBuilder.as(LogicalTypeAnnotation.timeType(time.isAdjustedToUTC, convertTimeUnit(time.unit)));
+                }
+                if (element.getLogicalType().isSetDECIMAL()) {
+                    DecimalType decimal = type.getDECIMAL();
+                    typeBuilder.as(LogicalTypeAnnotation.decimalType(decimal.scale, decimal.precision));
+                }
+                if (element.getLogicalType().isSetINTEGER()) {
+                    IntType integer = type.getINTEGER();
+                    typeBuilder.as(LogicalTypeAnnotation.intType(integer.bitWidth, integer.isSigned));
+                }
+                if (element.getLogicalType().isSetTIMESTAMP()) {
+                    TimestampType timestamp = type.getTIMESTAMP();
+                    typeBuilder.as(LogicalTypeAnnotation.timestampType(timestamp.isAdjustedToUTC, convertTimeUnit(timestamp.unit)));
+                }
+            }
```

Comment:
I only had to include these four because they are the ones that include logical type parameters that need to persist through the conversion process. This is an adaptation of the `getLogicalTypeAnnotation` method from the parquet metadata converter. The actual method does not work since it uses a switch statement on the wrong type, or a type that can't be used for some reason...

Reply:
Can you explain this a bit more?  I don't understand why we need this corresponding reader change.

Reply:
This is because we were losing the logical type information upon reading the parquet schema. The test will not pass without this logic in place. It's fine if we don't need logical type info for storage in certain databases/formats but within the parquet module it is important to at least read it so that choice can be made within other connectors

Reply:
Do you feel that there are sufficient tests in place to ensure that the type conversion logic works with the Parquet built in converter?

Reply:
Which converter are you referring to? I'm not sure what tests Apache has in place for their libraries...

Reply:
We've deleted the `MessageTypeConverter`, and I'm just wondering if the unit test coverage is sufficient to prove that using the built in converter from Parquet gives us 1:1 output otherwise.

Reply:
Let's work on some of those tests, the scope of the change makes me nervous that the existing unit test coverage isn't sufficient.

Reply:
I don't think we should special-case for only certain types when reading logical types. If a logical type is set, we should call a method which does the logical type conversion, and then sets the logical type in the descriptor. If the logical type isn't known during conversion, then throw an exception.

See here for an example: https://github.com/trinodb/trino/blob/0d5576a67a855bf95123acaf3ea836d1963861ef/lib/trino-parquet/src/main/java/io/trino/parquet/reader/MetadataReader.java#L221-L227

Reply:
I don't think this is necessary, the logical type annotations are already of the proper type.  I just added the date type which isn't included in the logic here, to `testWriteParquetFileWithLogicalTypes` and it is still passing. The only purpose of this logic is to make sure that the parameters are correct, so it only needs to check the logical types that take parameters. Adding conditions for each logical type would be overkill, and confusing since most of the conditions wouldn't actually result in any additional functionality. 

That being said, I can extract it into a method named something like `getLogicalTypeAnnotationWithParameters` to make the intentions a little more clear here.

Reply:
For completeness' sake, I do think that we should exhaustively convert all logical types so we don't have to touch this code afterwards. I agree it doesn't actually make much difference except for parameterized types or other logical types which are stored in the same physical format (e.g. long decimal and UUID). I would prefer to do the exhaustive conversion.

On another note, is this the error you encountered with the other approach?

```
java.lang.NoSuchMethodError: org.apache.parquet.format.LogicalType.getSetField()Lcom/facebook/presto/hive/$internal/parquet/org/apache/thrift/TFieldIdEnum;
	at org.apache.parquet.format.converter.ParquetMetadataConverter.getLogicalTypeAnnotation(ParquetMetadataConverter.java:1084)
	at org.apache.parquet.format.converter.ParquetMetadataConverter.buildChildren(ParquetMetadataConverter.java:1715)
	at org.apache.parquet.format.converter.ParquetMetadataConverter.fromParquetSchema(ParquetMetadataConverter.java:1670)
	at org.apache.parquet.format.converter.ParquetMetadataConverter.fromParquetMetadata(ParquetMetadataConverter.java:1526)
	at org.apache.parquet.format.converter.ParquetMetadataConverter.readParquetMetadata(ParquetMetadataConverter.java:1490)
	at org.apache.parquet.hadoop.ParquetFileReader.readFooter(ParquetFileReader.java:591)
	at org.apache.parquet.hadoop.ParquetFileReader.<init>(ParquetFileReader.java:799)
	at org.apache.parquet.hadoop.ParquetFileReader.open(ParquetFileReader.java:654)
	at org.apache.iceberg.parquet.ParquetUtil.fileMetrics(ParquetUtil.java:80)
	at org.apache.iceberg.parquet.ParquetUtil.fileMetrics(ParquetUtil.java:75)
	at com.facebook.presto.iceberg.IcebergParquetFileWriter.lambda$getMetrics$0(IcebergParquetFileWriter.java:77)
	at com.facebook.presto.hive.authentication.NoHdfsAuthentication.doAs(NoHdfsAuthentication.java:23)
```

Reply:
Yes I believe that was the error I got, the `getSetField` method does not exist. 

I can update it to an exhaustive list.